client/blocks/reader-post-card/index.jsx

```diff
@@ -27,6 +28,142 @@ import { getDiscoverBlogName,
 } from 'reader/discover/helper';
 import DiscoverFollowButton from 'reader/discover/follow-button';
 
+function trackable( TrackedComponent ) {
+	return class Trackable extends React.Component {
+		constructor( props ) {
+			super( props );
+			this.state = {
+				isOnScreen: false,
+				user: false,
+				readPixelKey: false,
+				readPixelInterval: false
+			};
+		}
+
+		defaultProps = {
+			onAppear: noop,
+			onHeartbeat: noop,
+			onLeave: noop
+		}
+
+		componentDidMount() {
+			window.addEventListener( 'scroll', this.checkOnScreen );
+			window.addEventListener( 'resize', this.checkOnScreen );
+			document.addEventListener( 'visibilitychange', this.checkOnScreen );
+			this.checkOnScreen();
+		}
+
+		componentWillUnmount() {
+			window.removeEventListener( 'scroll', this.checkOnScreen );
+			window.removeEventListener( 'resize', this.checkOnScreen );
+			document.removeEventListener( 'visibilitychange', this.checkOnScreen );
+		}
+
+		checkOnScreen = throttle( () => {
+			if ( ! this.nodeRef ) {
+				if ( this.state.readPixelInterval ) {
+					window.clearInterval( this.state.readPixelInterval );
+				}
+				return;
+			}
+
+			const rect = this.nodeRef.getBoundingClientRect();
+			const html = document.documentElement;
+			const windowHeight = window.innerHeight || html.clientHeight;
+
+			function documentNotHidden() {
+				if ( typeof document.hidden === 'undefined' ) {
+					return true;
+				}
+				return ! document.hidden;
+			}
+
+			function entirelyOnScreen() {
+				return rect.top >= 0 &&
+					rect.bottom <= windowHeight;
+			}
+
+			function biggerThanScreen() {
+				return rect.top < 0 &&
+					rect.bottom >= windowHeight;
+			}
+
+			function partiallyOnScreen() {
+				return ( rect.top >= 0 && rect.top <= windowHeight * .75 ) ||
+					( rect.bottom <= windowHeight && rect.bottom >= windowHeight * .25 );
+			}
+
+			function generateKey() {
+				const randomBytesLength = 18; // 18 * 4/3 = 24 (base64 encoded chars)
+				let i, randomBytes = [];
+
+				if ( window.crypto && window.crypto.getRandomValues ) {
+					randomBytes = new Uint8Array( randomBytesLength );
+					window.crypto.getRandomValues( randomBytes );
+				} else {
+					for ( i = 0; i < randomBytesLength; ++i ) {
+						randomBytes[ i ] = Math.floor( Math.random() * 256 );
+					}
+				}
+
+				return window
+					.btoa( String.fromCharCode.apply( String, randomBytes ) )
+					.replace( /\+/g, '-' )
+					.replace( /\//g, '_' );
+			}
+
+			if ( documentNotHidden() && ( entirelyOnScreen() || biggerThanScreen() || partiallyOnScreen() ) ) {
+				if ( ! this.state.isOnScreen ) {
+					const self = this;
+					const key = generateKey();
+					const user = userModule().get();
+					self.setState( {
+						isOnScreen: performance.now(),
+						user: user,
+						readPixelKey: key,
+						readPixelInterval: window.setInterval( function() {
```

Comment:
this should probably be a method on the class instead of a function set into state